File: source/python/bst_inorder.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def insert(root, value):
    if root is None:
        return Node(value)
    if value < root.value:
        root.left = insert(root.left, value)
    elif value > root.value:
        root.right = insert(root.right, value)
    return root

def inorder(root, result):
    if root is None:
        return
    inorder(root.left, result)
    result.append(root.value)
    inorder(root.right, result)

root = None
for v in [50, 30, 70, 20, 40, 60, 80]:
    root = insert(root, v)

result = []
inorder(root, result)
print(f"inorder: {' '.join(str(x) for x in result)}")

File: source/rust/bst_inorder.rs
type Tree = Option<Box<Node>>;

struct Node {
    value: i32,
    left: Tree,
    right: Tree,
}

fn insert(root: Tree, value: i32) -> Tree {
    match root {
        None => Some(Box::new(Node { value, left: None, right: None })),
        Some(mut node) => {
            if value < node.value {
                node.left = insert(node.left, value);
            } else if value > node.value {
                node.right = insert(node.right, value);
            }
            Some(node)
        }
    }
}

fn inorder(root: &Tree, result: &mut Vec<i32>) {
    if let Some(node) = root {
        inorder(&node.left, result);
        result.push(node.value);
        inorder(&node.right, result);
    }
}

fn main() {
    let mut root: Tree = None;
    for &v in &[50, 30, 70, 20, 40, 60, 80] {
        root = insert(root, v);
    }
    let mut result = Vec::new();
    inorder(&root, &mut result);
    let s: Vec<String> = result.iter().map(|x| x.to_string()).collect();
    println!("inorder: {}", s.join(" "));
}
